lib/charms/alertmanager_k8s/v0/alertmanager_dispatch.py

```diff
@@ -0,0 +1,257 @@
+    charm's metadata.yaml file:
+
+    ```yaml
+    provides:
+      alerting:
+        interface: alertmanager_dispatch
+    ```
+
+    A typical example of importing this library might be
+
+    ```python
+    from charms.alertmanager_k8s.v0.alertmanager_dispatch import AlertmanagerProvider
+    ```
+
+    In your charm's `__init__` method:
+
+    ```python
+    self.alertmanager_provider = AlertmanagerProvider(self, self._relation_name, self._api_port)
+    ```
+
+    Then inform consumers on any update to alertmanager cluster data via
+
+    ```python
+    self.alertmanager_provider.update_relation_data()
+    ```
+
+    This provider auto-registers relation events on behalf of the main Alertmanager charm.
+
+    Arguments:
+            charm (CharmBase): consumer charm
+            relation_name (str): relation name (not interface name)
+            api_port (int): alertmanager server's api port; this is needed here to avoid accessing
+                            charm constructs directly
+
+    Attributes:
+            charm (CharmBase): the Alertmanager charm
+    """
+
+    def __init__(self, charm, relation_name: str, api_port: int = 9093):
+        super().__init__(charm, relation_name)
+        self.charm = charm
+
+        self._api_port = api_port
+
+        events = self.charm.on[self.name]
+
+        # No need to observe `relation_departed` or `relation_broken`: data bags are auto-updated
+        # so both events are address on the consumer side.
+        self.framework.observe(events.relation_joined, self._on_relation_joined)
+
+    @property
+    def api_port(self):
+        """Get the API port number to use for alertmanager."""
+        return self._api_port
+
+    def _on_relation_joined(self, event: RelationJoinedEvent):
+        """This hook stores the public address of the newly-joined "alerting" relation.
+
+        This is needed for consumers such as prometheus, which should be aware of all alertmanager
+        instances.
+        """
+        self.update_relation_data(event)
+
+    def _generate_relation_data(self, relation: Relation):
+        """Helper function to generate relation data in the correct format."""
+        public_address = (
+            f"{self.charm.model.get_binding(relation).network.bind_address}:{self.api_port}"
+        )
+        return {"public_address": public_address}
+
+    def update_relation_data(self, event: RelationEvent = None):
+        """Helper function for updating relation data bags.
+
+        This function can be used in two different ways:
+        - update relation data bag of a given event (e.g. a newly joined relation);
+        - update relation data for all relations
+
+        Args:
+            event: The event whose data bag needs to be updated. If it is None, update data bags of
+            all relations.
+        """
+        if event is None:
+            # update all existing relation data
```

Comment:
I am not sure why you need to do these two checks `event is None` and `self.name in self.charm.model.relations`. Could you not just iterate over relations of a specific name as in `for relation in self.charm.model.relations[<relation-name>]` ?

Reply:
@sed-i 👆

Reply:
iirc the `if self.name in relations` is there because `self.charm.model.relations[self.name]` on its own occasionally raised KeyError. Can't think of another reason I would have had it there for.